pkg/operator/observe_config.go

```diff
@@ -0,0 +1,151 @@
+package operator
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	corev1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorconfigclientv1alpha1 "github.com/openshift/cluster-openshift-controller-manager-operator/pkg/generated/clientset/versioned/typed/openshiftcontrollermanager/v1alpha1"
+	operatorconfiginformerv1alpha1 "github.com/openshift/cluster-openshift-controller-manager-operator/pkg/generated/informers/externalversions/openshiftcontrollermanager/v1alpha1"
+)
+
+type ConfigObserver struct {
+	operatorConfigClient operatorconfigclientv1alpha1.OpenShiftControllerManagerOperatorConfigInterface
+
+	kubeClient kubernetes.Interface
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+
+	rateLimiter flowcontrol.RateLimiter
+}
+
+func NewConfigObserver(
+	operatorConfigInformer operatorconfiginformerv1alpha1.OpenShiftControllerManagerOperatorConfigInformer,
+	kubeAPIServerNamespacedKubeInformers kubeinformers.SharedInformerFactory,
+	etcdNamespacedKubeInformers kubeinformers.SharedInformerFactory,
+	operatorConfigClient operatorconfigclientv1alpha1.OpenshiftcontrollermanagerV1alpha1Interface,
+	kubeClient kubernetes.Interface,
+) *ConfigObserver {
+	c := &ConfigObserver{
+		operatorConfigClient: operatorConfigClient.OpenShiftControllerManagerOperatorConfigs(),
+		kubeClient:           kubeClient,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
+
+		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+	}
+
+	operatorConfigInformer.Informer().AddEventHandler(c.eventHandler())
+	kubeAPIServerNamespacedKubeInformers.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+	etcdNamespacedKubeInformers.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+// sync reacts to a change in prereqs by finding information that is required to match another value in the cluster. This
+// must be information that is logically "owned" by another component.
+// TODO(juanvallejo): update this to actually watch this operator's config
+func (c ConfigObserver) sync() error {
+	return nil
```

Comment:
bring back enough to read the current value and serialize and write the empty map in.